runtime/lua/vim/lsp/util.lua

```diff
@@ -159,8 +159,10 @@ end
 function M.apply_text_document_edit(text_document_edit)
   local text_document = text_document_edit.textDocument
   local bufnr = vim.uri_to_bufnr(text_document.uri)
-  -- `VersionedTextDocumentIdentifier`s version may be nil https://microsoft.github.io/language-server-protocol/specification#versionedTextDocumentIdentifier
-  if text_document.version ~= vim.NIL and M.buf_versions[bufnr] > text_document.version then
+  -- `VersionedTextDocumentIdentifier`s version may be null https://microsoft.github.io/language-server-protocol/specification#versionedTextDocumentIdentifier
+  -- Comparing against vim.NIL as that's Json null's decoded value
+  -- We also apply change if buffer version is nil, possible when buffer not attached to lsp client yet
```

Comment:
Is this case, for example, "textDocument/rename"?

Reply:
Good catch! I'll update the PR tonight adding rename coverage that would have caught the issue. I think master is broken for the rename case after #12237 was merged.

Reply:
note also #12223 (including a test for non-ASCII edits)